crates/oxc_linter/src/rules/typescript/prefer_for_of.rs

```diff
@@ -0,0 +1,296 @@
+            BindingPatternKind::BindingIdentifier(id) => (&id.name, id.symbol_id.get()),
+            _ => return,
+        };
+
+        if !matches!(&decl.init,
+            Some(Expression::NumericLiteral(literal)) if literal.value == 0f64
+        ) {
+            return;
+        }
+
+        let Some(Expression::BinaryExpression(test_expr)) = &for_stmt.test else { return };
+
+        if !matches!((&test_expr.left, test_expr.operator),
+            (Expression::Identifier(id), BinaryOperator::LessThan) if id.name == var_name
+        ) {
+            return;
+        }
+
+        let array_name = {
+            let Expression::MemberExpression(mem_expr) = &test_expr.right else { return };
+            if !matches!(mem_expr.static_property_name(), Some(prop_name) if prop_name == "length")
+            {
+                return;
+            }
+
+            match &mem_expr.object() {
+                Expression::Identifier(id) => id.name.as_str(),
+                Expression::MemberExpression(m_expr) => match m_expr.static_property_name() {
+                    Some(array_name) => array_name,
+                    None => return,
+                },
+                _ => return,
+            }
+        };
+
+        let Some(update_expr) = &for_stmt.update else { return };
+        if !update_expr.is_increment_of(var_name) {
+            return;
+        }
+
+        let nodes = ctx.nodes();
+        let test_and_update_span = test_expr.span.merge(&update_expr.span());
+        let body_span = for_stmt.body.span();
+
+        let Some(var_symbol_id) = var_symbol_id else { return };
+        if ctx.semantic().symbol_references(var_symbol_id).any(|reference| {
+            let ref_id = reference.node_id();
+
+            let symbol_span = nodes.get_node(ref_id).kind().span();
+            if test_and_update_span.contains(symbol_span) || !body_span.contains(symbol_span) {
+                return false;
+            }
+
+            let Some(ref_parent) = nodes.parent_node(ref_id) else { return true };
+            if matches!(nodes.parent_node(ref_parent.id()), Some(ref_grand_parent)
+                if is_assignee(ref_grand_parent))
+            {
+                return true;
+            }
+
+            let parent_kind = ref_parent.kind();
+            let AstKind::MemberExpression(mem_expr) = parent_kind else { return true };
+            let MemberExpression::ComputedMemberExpression(com_mem_expr) = &mem_expr else {
+                return true;
+            };
+
+            match &com_mem_expr.object {
+                Expression::Identifier(id) => id.name.as_str() != array_name,
+                Expression::MemberExpression(mem_expr) => {
+                    matches!(&**mem_expr,
+                        MemberExpression::StaticMemberExpression(st_mem_expr)
+                        if st_mem_expr.property.name != array_name
+                    )
+                }
+                _ => true,
+            }
```

Comment:
Can you explain why we need this part? Didn't find the corresponding part in [prefer-for-of](https://github.com/typescript-eslint/typescript-eslint/blob/04e32d6834c19160fb4f44f812a3e64671fca893/packages/eslint-plugin/src/rules/prefer-for-of.ts#L153-L154).🤔

And I didn't find our logic of checking [`ThisExpression`](https://github.com/typescript-eslint/typescript-eslint/blob/04e32d6834c19160fb4f44f812a3e64671fca893/packages/eslint-plugin/src/rules/prefer-for-of.ts#L117).

Reply:
There is an `isIndexOnlyUsedWithArray` function. In the original, they are checking it as "if ALL of the references are used only with an array (from the `test` expression), then suggest 'prefer-for-of'".

The ported version uses reversed logic: "if ANY of the references are used not with the array (from the `test` expression), then skip the rule". In our case, returning `true` means that the reference "is bad".

```ts
const arrayText = context.sourceCode.getText(arrayExpression);
return indexVar.references.every(reference => {
  const id = reference.identifier;
  const node = id.parent;

  return (
    !contains(body, id) || (
        node.type === AST_NODE_TYPES.MemberExpression &&
        node.object.type !== AST_NODE_TYPES.ThisExpression &&
        node.property === id &&
        context.sourceCode.getText(node.object) === arrayText &&
        !isAssignee(node)
    )
  );
});

```

```rust
let parent_kind = ref_parent.kind();
let AstKind::MemberExpression(mem_expr) = parent_kind else { return true };
let MemberExpression::ComputedMemberExpression(com_mem_expr) = &mem_expr else {
    return true;
};

match &com_mem_expr.object {
    Expression::Identifier(id) => id.name.as_str() != array_name,
    Expression::MemberExpression(mem_expr) => {
        matches!(&**mem_expr,
            MemberExpression::StaticMemberExpression(st_mem_expr)
            if st_mem_expr.property.name != array_name
        )
    }

    // Probably a better approach...
    // Expression::MemberExpression(mem_expr) => match mem_expr.static_property_name() {
    //     Some(prop_name) => prop_name != array_name,
    //     None => true,
    // },

    // node.object.type !== AST_NODE_TYPES.ThisExpression from the original goes here
    _ => true,
}

```

However, I may be mistaken somewhere, so please don't take my words at face value.